common/djangoapps/student/roles.py

```diff
@@ -60,21 +62,47 @@ def strict_role_checking():
     ACCESS_ROLES_INHERITANCE.update(OLD_ACCESS_ROLES_INHERITANCE)
 
 
+def get_role_cache_course_key(course_id=None):
+    """
+    Get the cache key for the course key.
+    """
+    default = ROLE_CACHE_UNGROUPED_COURSES_KEY
+    if not course_id:
+        return default
+    if not getattr(course_id, 'html_id', False):
+        return default
+    html_id = course_id.html_id()
+    if not html_id:
+        return default
+
+    return html_id
+
+
 class BulkRoleCache:  # lint-amnesty, pylint: disable=missing-class-docstring
     CACHE_NAMESPACE = "student.roles.BulkRoleCache"
     CACHE_KEY = 'roles_by_user'
 
     @classmethod
     def prefetch(cls, users):  # lint-amnesty, pylint: disable=missing-function-docstring
-        roles_by_user = defaultdict(set)
+        roles_by_user = {}
         get_cache(cls.CACHE_NAMESPACE)[cls.CACHE_KEY] = roles_by_user
 
         for role in CourseAccessRole.objects.filter(user__in=users).select_related('user'):
```

Comment:
Question: I get that you want to grab the access role data in one pass, but why put them into two dicts with differing key types (course keys vs. user_ids), as opposed to two different dicts? Backwards compatibility concerns?

Reply:
I'm having a hard time understanding the question. Let me try to break it down.
The current idea I'm working off of is that this data structure is inefficient for checking if a specific user has a role for a specific course because one user can have lots of roles and you'll have to iterate over all of them.
So I'm changing the previous data structure to the new data structure as I'll show here in pseudocode:

previous:
```
roles_in_cache = { 'user_1': set(role1, role2, role3, ...), 'user2': set(roleX, roleY, roleZ, ...), ... }
```

new:
```
roles_in_cache = { 'user_1': { 'course_1': set(role1), 'course_2': set(role2) }, 'user_2': {'course_1': set(roleX, roleY), 'course_2': set(roleZ) }
```

The previous version was already a dict that groups roles by users so that there's no iteration over all roles that exist. However, it's still not so good because you need to iterate over all roles one user has, since that's a set, and some users have thousands of courses and libraries and at least one role for each.
The new version just further optimizes this by keeping the dict that groups roles by users and inside of that nesting a dict that groups roles for a specific user by course as well.

The result is that instead of needing to iterate over all roles that a user_1 has, you only need to iterate over the roles that user_1 has that belong to the specific course you're looking for.
This works so well because while a user may have thousands of courses and each has its own role, if you group by course, the user will usually just have something like 1-3 roles for that course.

Now I'm not sure I understand your suggestion about two dicts. But let me interpret that you're suggesting that we have one cache dict that groups all roles by user, and another one that groups all roles by course. Now that would not improve the situation or even make it worse, because now instead of looking up the roles for user_1 for course_1, which should be at most two or three, we are looking up in a set of all roles for course_1, which again may be many thousands because many users could have that role, and we'll need to search for the user we want again.

Reply:
@jesperhodge I believe that when @ormsbee said

> but why put them into two dicts with differing key types (course keys vs. user_ids) ... ?

he was specifically referring to this code:
```python
            if not roles_by_user.get(course_id):
                roles_by_user[course_id] = set()
            roles_by_user[course_id].add(role)
```

which you've since clarified was mistaken.

Reply:
Yup, that was all I meant. Thank you.